ESPS/general/src/mux/mux.1
.\" Copyright (c) 1986-1990 Entropic Speech, Inc.
.\" Copyright (c) 1991 Entropic Research Laboratory, Inc. All rights reserved.
.\" @(#)mux.1	1.4 03 Apr 1997 ESI/ERL
.ds ]W (c) 1991 ERL, Inc.
.TH MUX 1\-ESPS 03 Apr 1997
.if t .ds - \(em\h'-0.2m'\(em
.if n .ds - ---
.SH NAME

.nf
mux \- multiplex sampled-data files into a single multichannel or complex file
.fi
.SH SYNOPSIS
.B mux
[
.BR \- { prs } " \fIrange\fP"
] .\ .\ . [
.BI \-x " debug_level"
] [
.B \-J
] [
.BI \-P " param_file"
]
.I input1.fsd
[
.I "input2.fsd .\ .\ ."
]
.I output.fsd
.SH DESCRIPTION
.PP
The
.I mux
(``multiplex'') program combines its input sampled-data FEA_SD(5\-ESPS) files,
or equal-length portions of them,
into a single multichannel output sampled-data file,
possibly also combining real channels in pairs to form complex channels.
.PP
Normally the number of output channels
is the total number of channels in all the input files.
Each output record contains the data from the
.I samples
field in one record of each input file.
This is organized as a single vector containing one sample value
from each channel of each input file.
Within the vector,
the data from the first input file comes first,
followed by the data from the second, and so on,
in the order of the file names on the command line.
The channels of any one input file
keep the same relative ordering in the output file
that they had in the input file.
.PP
When the
.B \-J
option is used, the number of output channels is only half the normal number;
input channels of a real data type are combined in pairs
into single output channels of the corresponding complex type.
.PP
The input files must be consistent in data type and sampling frequency;
the output file has the same data type
(unless
.B \-J
is used)
and the same sampling frequency as the input files.
Any fields other than
.I samples
in the input are ignored.
.PP
By default, the first output record contains data from the first
record in each input file, and in general the
.IR n th
output record contains data from the
.IR n th
record in each input file;
but a later starting point in each input file can be chosen with the
.BR \-p ,
.BR \-r ,
or
.B \-s
option.
By default, records are processed until an input file runs out of data, 
but a shorter range of data can be chosen with
.BR \-p ,
.BR \-r ,
or
.BR \-s .
.PP
If ``\-'' is written for an input file, the standard input is used.
At most one input file may be standard input.
Names of disk files, however, may be repeated (duplicating channels).
Since different 
.BR \-p ,
.BR \-r ,
and
.BR \-s
options may apply to each instance of a repeated input file name,
it is possible
to align and juxtapose different portions of a single input file.
If ``\-'' is written for the output, the standard output is used.
.SH OPTIONS
.PP
The following options are supported:
.TP
.BI \-p " range"
For this program
.B \-p
and
.B \-r
are synonymous.
See
.B \-r
for the interpretation and the format of the argument.
.TP
.BI \-r " first" : last
.TP
.BI \-r " first" :+ incr
Determines the range of points (records) to be taken from an input file.
In the first form, a pair of unsigned integers
gives the numbers of the first and last records of the range.
(Counting starts with 1 for the first record in the file.)
If
.I first
is omitted, 1 is used.
If
.I last 
is omitted, the last record in the file is used.
The second form is equivalent to the first with 
.I "last = first + incr".
.IP
This option and the
.BR \-p
and
.B \-s
options may be repeated up to a maximum total number, for all three kinds,
of the number of input files.
The first
.BR \-p ,
.BR \-r ,
or
.B \-s
option applies to the first input file, the second to the second,
and so on.
If there are fewer
.BR \-p ,
.BR \-r ,
and
.B \-s
options than input files,
the last such option applies to all the remaining input files.
In particular, if there is only one
.BR \-p ,
.BR \-r ,
or
.B \-s
option, it applies to all the input files.
.IP
If two options disagree as to the number of records to be processed,
the smaller number applies.
In fact
.I mux
stops processing as soon as it encounters either the end of a
specified range or the actual end of file in any input file.
Certain inconsistencies in these various stopping criteria
will evoke warning messages; see the Diagnostics section for details.
.TP
.BI \-s " start" : end
.TP
.BI \-s " start" :+ incr
Determines the range in seconds of the data to be taken from an input file.
In the first form, a pair of floating-point numbers give
the beginning time and ending time of the range.
The second form is equivalent to the first with 
.I "last = first + incr."
Each sample has a time given by
.IR s " + (" r "\-1)/" f,
where
.I s
is the value of the generic header item "start_time",
.I r
is the record number, and
.I f
is the sampling frequency, given by the generic header item "record_freq".
This time may depend on the channel number,
since the "start_time" item may be a vector with a component per channel;
for present purposes the value for the first channel (number 0) is used.
The range selected by the
.B \-s
option consists of the records for which the time is less than
.I end
but not less than
.I start.
.IP
This option and the
.B \-p
and
.B \-r
options may be repeated to supply different ranges for different input files.
See the
.B \-r
option for details.
.TP
.BI \-x " debug_level"
If
.I debug_level
is positive, the program prints debugging messages as it progresses\*-\c
the higher the number, the more messages.
The default level is 0, for no debugging output.
.TP
.B \-J
Join pairs of input channels to form single complex output channels.
The total number of channels in the input files must be even,
and the output file has half that number of channels.
The input channels are taken in the usual order and grouped in pairs
to form the real and imaginary parts of the output channels.
The pairing is without regard to
whether two input channels come from the same input file or consecutive files.
The last channel of a file,
if not paired with the previous input channel,
is paired with the first channel of the next input file.
.IP
The input files must all have the same real data type:
DOUBLE, FLOAT, LONG, SHORT, or BYTE.
(See
.IR FEA (5-ESPS)
for an explanation of these type codes.)
The output file has the corresponding complex data type:
DOUBLE_CPLX, FLOAT_CPLX, LONG_CPLX, SHORT_CPLX, or BYTE_CPLX.
.IP
If two channels with with different time alignments
are combined into one complex channel,
time-alignment information may be lost.
A warning message is printed in that case.
See the discussion of the "start_time" generic header item
in the section on ESPS Headers.
.TP
.BI \-P " param_file"
The name of the parameter file.
The default name is ``params''.
.SH "ESPS PARAMETERS"
.PP
The parameter file is not required to be present, as there are default
values for all parameters.
If the file exists, the following parameters may be read
if they are not determined by command-line options.
.TP
.I "start \- integer array"
The starting record number in each input file.
The array elements are matched with input files in order.
If there are more input files,
the last array element applies to the unmatched file.
If there are more array elements, the unmatched ones are ignored.
This parameter is not read if the
.BR \-p ,
.BR \-r ,
or
.B \-s
option is specified.
The default is all 1's, meaning the beginning of each input file.
.TP
.I "nan \- integer"
The number of records to process in each input file.
A value of 0 (the default) means
continue processing until the end of an input file is reached.
This parameter is not read if the
.BR \-p ,
.BR \-r ,
or
.B \-s
option is specified.
.TP
.I "make_complex \- string"
A value of "YES" or "yes" means
join pairs of real channels to form complex channels as if the
.B \-J
option is in force.
A value of "NO" or "no"
means make a separate output channel for each input channel as usual.
No other values are allowed.
This parameter is not read if the
.B \-J
option is specified.
The default value is "NO".
.SH "ESPS COMMON"
.PP
The ESPS Common file is not read.
.PP
If Common processing is enabled, and the output file is not standard output,
the program writes the Common parameters
.I prog,
.I filename,
.I start,
and
.I nan
to record the program's name, the name of the output file,
the starting record number of the output file (always 1),
and the number of points in the output file.
.PP
ESPS Common processing may be disabled
by setting the environment variable USE_ESPS_COMMON to
.I off.
The default ESPS Common file is
.I espscom
in the user's home directory.
This may be overidden
by setting the environment variable ESPSCOM to the desired path.
User feedback of Common processing is determined
by the environment variable ESPS_VERBOSE, with 0 causing no feedback
and increasing levels causing increasingly detailed feedback.
If ESPS_VERBOSE is not defined, a default value of 3 is assumed.
.SH "ESPS HEADERS"
.PP
The output header is a new FEA_SD file header,
with appropriate items copied from the input headers.
.PP
The generic header item "record_freq",
which must have the same value in all input files,
is copied into the output header.
.PP
The generic header item "start_time"
records the starting time for each output channel.
It is a single number if all output channels have the same starting time;
otherwise it is a vector with one element per channel.
The starting time for a channel is its starting time in the input file
plus an offset in case the data taken from the input file
do not start with the first record.
The offset is given by
.RI ( r \-1)/ f
where
.I r
is the starting record number in the input file and
.I f
is the sampling frequency given by the "record_freq" header item.
The
.B \-J
option can create complex channels
whose real and imaginary parts have inconsistent starting times.
When that happens, a warning message is printed, and the starting time
for the real part is recorded in the "start_time" header item.
.PP
If every input file has a "max_value" header item,
then the output file
has a "max_value" header item containing the same information.
.SH EXAMPLES
.PP
Multiplex data from three input files to produce an output file
.I xxx.
Processing begins with the sampled data
in the first record in each input file.
The output file has the same length as the shortest input file.
.IP
.I "mux aaa bbb ccc xxx"
.PP
Start at time 0.5 in each input file
and process 0.5 seconds of data from each.
(Suppose the sampling frequency is 8000 Hz,
and the start times in the three input files are 0.0, 0.0, and 0.5.
Then the starting record numbers are 4001, 4001, and 1, respectively.
The start time in the output file is 0.5 for all channels.)
.IP
.I "mux \-s0.5:1.0 aaa bbb ccc xxx"
.PP
Start at time 0.5 in file
.I aaa
and with the first record in the other two input files.
(With the assumptions of the previous example,
the starting record numbers in the three input files
are 4001, 1, and 1, respectively.
The start times in the output file header are 0.5 for data from files
.I aaa
and
.I ccc
and 0.0 for data from file
.IR bbb. )
.IP
.I "mux \-s0.5: \-p1: aaa bbb ccc xxx"
.PP
Juxtapose data from an input file with a test signal
and pass the result to another program.
.IP
.I "testsd \- | mux aaa \- \- | more_processing \-"
.PP
If
.I aaa
has two channels of real data,
this will convert it to a single-channel file
.I zzz
of complex data.
.IP
.I "mux \-J aaa zzz"
.PP
If
.I aaa
and
.I bbb
are single-channel files of real data,
this will join them into a single-channel file of complex data.
.IP
.I "mux \-J aaa bbb zzz"
.PP
Multiplex a portion of a file with a later portion of the same file.
.IP
.I "mux \-p1001:2000 \-p2501: aaa aaa xxx"
.SH "SEE ALSO"
.PP
demux(1-ESPS), copysps(1-ESPS), addgen(1-ESPS), 
.br
FEA_SD(5-ESPS), FEA(5-ESPS)
.SH DIAGNOSTICS
.PP
The program exits with an error message if any of the following occur.
  The command line cannot be parsed.
  Fewer than two file names are specified (one in, one out).
  Fewer input file names are specified than
.BR \-p ,
.BR \-r ,
and
.B \-s
options.
  More than one input file name is ``\-''.
  An input file cannot be opened or is not an ESPS sampled-data file.
  The input files do not all have the same sampling frequency.
  The input files do not all have the same data type.
  The
.B \-J
option is specified with input files of a complex data type.
  The
.B \-J
option is specified, and the total number of input channels is odd.
  A starting record specified with a
.BR \-p ,
.BR \-r ,
or
.B \-s
option does not exist in all the files that the option applies to.
.fi
.PP
The program issues a warning message if the end of a range specified by a
.BR \-p ,
.BR \-r ,
or
.B \-s
option is not reached,
and the option argument (see the Options section) ends with an explicit
.I last,
.I end,
or
.BI + incr.
(This doesn't apply to option arguments that default to end-of-file
by omitting what follows the colon.)
The end of the range may fail to be reached
either because the end of an input file is reached first
or because another
.BR \-p ,
.BR \-r ,
or
.B \-s
option causes an earlier stop.
.PP
The program issues a warning message if processing for the
.B \-J
option joins two channels that are not properly time-aligned
(so that they would require conflicting entries
in the output "start_time" header item).
.SH BUGS
.PP
The \fB-s\fR option is not implemented in this version of the program.
.SH "AUTHOR"
Manual page by Rodney Johnson.
Program by Alan Parker.




ESPS/general/src/mux/mux.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1986-1990  Entropic Speech, Inc. 
 *    "Copyright (c) 1990-1991  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  Alan Parker
 * Checked by:
 * Revised by:  Rodney Johnson
 *
 * Multiplexes sampled-data files into a single multichannel or complex file.
 *
 */

static char *sccs_id = "@(#)mux.c	1.5	8/31/95	ESI/ERL";

#include <stdio.h>
#include <esps/esps.h>
#include <esps/fea.h>
#include <esps/feasd.h>
#include <esps/unix.h>


char	*Version = "1.5";
char	*Date = "8/31/95";

#define ERROR_EXIT(text) {(void) fprintf(stderr, "%s: %s - exiting\n", \
                ProgName, text); exit(1);}

#define Fprintf (void)fprintf

#define SYNTAX \
USAGE("mux [-{prs} range] ... [-x debug] [-J] [-P param_file] \n input1.fsd [input2.fsd ...] output.fsd")


#define INITIAL_INPUT_SIZE 20

/*
 * global variables
 */
char	*ProgName = "mux";
int	debug_level = 0;                /* debug level */
long	get_chans();
int	typesiz();
char	*get_cmd_line();
void	lrange_switch();
int	symsize();
int	typesiz();


main(argc, argv)
int	argc;
char	**argv;

{
	char	*param_file = NULL;  /*parameter file name*/
	FILE    *out_fd = stdout;	 /* output stream */
	char	*ofile;
	struct header	*oh;
	extern int	optind;
	extern char	*optarg;
	int	ch;

	char	**r_range;
	char	**s_range;
	int	range_size = INITIAL_INPUT_SIZE;
	int	range_count = 0;
	int	*ps_point=NULL, *pnan=NULL;
	

	FILE	**in_fd;
	struct header **ih;
	char	**in_files;
	int	input_size = 0;
	long	*e_point;
	long	*s_point;
	double  *start_time_file;
	double  *start_times;
	int	start_time_chan = 0;
	long	*n_read;

	int	havestdin = 0;
	int	nfiles = 0;
	int	Jflag = 0;
	double	first_sf;
	int	i,j;
	int	first_type;
	long	num_channels;
	int	total_channels = 0;
	int	out_type;
	double	*max_value_file;
	double	*max_values;
	int	max_value_chan = 0;
	int	genhd_size;
	int	no_max_value = NO, multi_start_time;
	double	start_time_tmp, start_time_offset;
	struct	feasd **in_rec;
	struct	feasd *out_rec;
	char	*in_data, *out_data;
	int	files_done=0;
	int	prev_num_chans;
	long	output_count=0;
	int	start_size=0, nan_size=0;
	char	*cptr;


	r_range = (char **)malloc(sizeof(char *) * INITIAL_INPUT_SIZE);
	s_range = (char **)malloc(sizeof(char *) * INITIAL_INPUT_SIZE);
	in_files = (char **)malloc(sizeof(char *) * INITIAL_INPUT_SIZE);
	in_fd = (FILE * *)malloc(sizeof(FILE * ) * INITIAL_INPUT_SIZE);
	ih = (struct header **)malloc(sizeof(struct header *) * 
		INITIAL_INPUT_SIZE);
	spsassert(r_range,"mux: malloc failed");
	spsassert(s_range,"mux: malloc failed");
	spsassert(in_files,"mux: malloc failed");
	spsassert(in_fd,"mux: malloc failed");
	spsassert(ih,"mux: malloc failed");


	while ((ch = getopt(argc, argv, "p:r:s:x:JP:")) != EOF)
		switch (ch)  {
		case 'x':
			debug_level = atoi(optarg);
			break;
		case 'J':
			Jflag = 1;
			break;
		case 'P':
			param_file = optarg;
			break;
		case 'r':
		case 'p':
			range_count++;
			if (range_count > range_size) {
				r_range = (char **)realloc((char *)r_range, 
				    sizeof(char *) * range_count);
				s_range = (char **)realloc((char *)s_range, 
				    sizeof(char *) * range_count);
				range_size++;
			}
			r_range[range_count-1] = optarg;
			s_range[range_count-1] = NULL;
			break;
		case 's':
			ERROR_EXIT("s option not done yet");
			range_count++;
			if (range_count > range_size) {
				r_range = (char **)realloc((char *)r_range, 
				    sizeof(char *) * range_count);
				s_range = (char **)realloc((char *)s_range, 
				    sizeof(char *) * range_count);
				range_size++;
			}
			s_range[range_count-1] = optarg;
			r_range[range_count-1] = NULL;
			break;
		default:
			SYNTAX;
		}

/* Determine and open output file
 */
	if (argc - optind <= 1) {
		Fprintf(stderr, "mux: Input and output filenames needed.\n");
		SYNTAX;
	}

	ofile = eopen("mux", argv[argc - 1], "w", NONE, NONE, &oh, &out_fd);
	if (debug_level) 
		Fprintf(stderr, "mux: output file is %s\n", ofile);
	oh = new_header(FT_FEA);



/* open the parameter file
 */

	(void) read_params(param_file, SC_NOCOMMON, (char *) NULL);

	if(!Jflag && symtype("make_complex") != ST_UNDEF){
		cptr = getsym_s("make_complex");
		if (!strcmp(cptr, "YES") || !strcmp(cptr, "yes"))
			Jflag = 1;
	}

	if(!range_count && symtype("start") != ST_UNDEF) {
		start_size = symsize("start");
		if (start_size != -1) {
			ps_point = (int *)
				calloc((unsigned)start_size, sizeof(int));
			(void)getsym_ia("start", ps_point, start_size);
		}
	}
	if(!range_count && symtype("nan") != ST_UNDEF) {
		nan_size = symsize("nan");
		if (nan_size != -1) {
			pnan = (int *)
				calloc((unsigned)nan_size, sizeof(int));
			(void)getsym_ia("nan", pnan, nan_size);
		}
	}


/*
 * Determine input files and check them -- start by getting the first 
 * one; at the same time, we set up to create the output header
 */
	if (argc - optind < 2) {
		Fprintf(stderr, "mux: no input files on command line\n");
		SYNTAX;
	}

	while (optind < argc - 1) {
		nfiles++;
		if (nfiles > input_size) {
			ih = (struct header **)realloc(ih, 
				sizeof(struct header *) * nfiles);
			in_fd = (FILE **)realloc((char *)in_fd, 
				sizeof(FILE *) * nfiles);
			in_files = (char **)realloc((char *)in_files, 
				sizeof(char *) * nfiles);
			input_size++;
		}
		in_files[nfiles-1] = eopen("mux", argv[optind], 
			"r", FT_FEA, FEA_SD, &ih[nfiles-1], &in_fd[nfiles-1]);

		if (in_fd[nfiles-1] == stdin) {
			if (havestdin) {
				ERROR_EXIT("only one stdin allowed");
			} else
				havestdin = 1;
		}
		add_source_file(oh, in_files[nfiles-1], ih[nfiles-1]);
		optind++;
	}
	add_comment (oh, get_cmd_line(argc, argv));
	(void) strcpy (oh ->common.prog, ProgName);
	(void) strcpy (oh ->common.vers, Version);
	(void) strcpy (oh ->common.progdate, Date);

	if (range_count) {
		if (nfiles > range_size) {
			r_range = (char **)realloc((char *)r_range, 
			    sizeof(char *) * nfiles);
			s_range = (char **)realloc((char *)s_range, 
			    sizeof(char *) * nfiles);
			spsassert(r_range && s_range, "mux: realloc failed");
			range_size = nfiles;
		}
		while (nfiles > range_count++) {
			r_range[range_count-1] = r_range[range_count-2];
			s_range[range_count-1] = s_range[range_count-2];
		}
	}
	else {
		if (start_size && nfiles > start_size){
			ps_point = (int *)
				realloc((char *)ps_point, nfiles*sizeof(int));
			while (start_size++ < nfiles) {
				ps_point[start_size-1]=ps_point[start_size-2];
			}
		}
		if (nan_size && nfiles > nan_size){
			pnan = (int *)
				realloc((char *)pnan, nfiles*sizeof(int));
			while (nan_size++ < nfiles) {
				pnan[nan_size-1]=pnan[nan_size-2];
			}
		}
	}

	if(debug_level) {

		for (i = 0; i < nfiles; i++) {
			Fprintf(stderr, "File: %s, range: %s - %s\n",
		    	in_files[i], r_range[i], s_range[i]);
		}
		Fprintf(stderr, "Jflag: %d\n", Jflag);
		Fprintf(stderr, "Param File: %s", param_file);
		Fprintf(stderr, " Output File: %s\n", ofile);
	}

/* check the input files for various problems
 */
	first_sf = get_genhd_val("record_freq", ih[0], (double)-1);
	first_type = get_fea_type("samples", ih[0]);

	s_point = (long *)calloc((unsigned)nfiles, sizeof(long));
	spsassert(s_point, "mux: calloc failed!");
	e_point = (long *)calloc((unsigned)nfiles, sizeof(long));
	spsassert(s_point, "mux: calloc failed!");
	n_read = (long *)calloc((unsigned)nfiles, sizeof(long));
	spsassert(n_read, "mux: calloc failed!");

	for (i = 0; i < nfiles; i++)
		total_channels += get_chans(ih[i]);

	if (Jflag && total_channels % 2) {
		ERROR_EXIT(
		 "J flag is used and total number of channels is odd.\n");
	}

	start_times = (double *) calloc((unsigned)total_channels, 
		sizeof(double));
	spsassert(start_times, "mux: calloc failed!");
	max_values = (double *) calloc((unsigned)total_channels, 
		sizeof(double));
	spsassert(max_values, "mux: calloc failed!");

	for (i = 0; i < nfiles; i++) {
		if (get_fea_type("samples", ih[i]) == UNDEF) {
			Fprintf(stderr,
		    	 "mux: Input file: %s has no field named samples!\n",
			 in_files[i]);
			exit(1);
		}
		if (Jflag && is_field_complex(ih[i], "samples")) {
			Fprintf(stderr,
			 "mux: J option is used and file %s has a complex field- exiting.\n",
			    in_files[i]);
			exit(1);
		}
		if (get_genhd_val("record_freq",ih[i],(double)0) != first_sf) {
			Fprintf(stderr,
			 "mux: all files must have same sampling freq\n");
			Fprintf(stderr,
			 "mux: freq. in first file is %g, freq., in %s is %g\n",
			    first_sf, in_files[i],
			    get_genhd_val("record_freq", ih[i], (double)0));
			exit(1);
		}
		if (get_fea_type("samples", ih[i]) != first_type) {
			Fprintf(stderr,
			 "mux: all files must have the same data type\n");
			Fprintf(stderr,
			 "mux: type of first file is %s, type of %s is %s.\n",
			    type_codes[first_type], in_files[i],
			    type_codes[get_fea_type("samples",ih[i])]);
			exit(1);
		}

		num_channels = get_chans(ih[i]);

		if (genhd_type("start_time", &genhd_size, ih[i]) != DOUBLE) {
			Fprintf(stderr,
			 "mux: header item \"start_time\" in file %s is undefined or not DOUBLE.\n", 
				in_files[i]);
			exit(1);
		}
		if (genhd_size != num_channels && genhd_size != 1) {
			Fprintf(stderr,
			 "mux: header item \"start_time\" in file %s has wrong size (%d).\n",
				in_files[i], genhd_size);
			exit(1);
		}
		start_time_file = get_genhd_d("start_time", ih[i]);

		s_point[i]=1;
		if(ih[i]->common.ndrec == -1) ih[i]->common.ndrec = LONG_MAX;
		e_point[i] = ih[i]->common.ndrec;	

		if (!range_count) {
			if (start_size) s_point[i] = ps_point[i];
			if (nan_size && pnan[i] != 0) 
				e_point[i] = s_point[i] + pnan[i] - 1;
		}
		else if (r_range[i])
			lrange_switch(r_range[i], s_point+i, e_point+i, 1);
		else { /* Handle s_range here. */ }

		if (e_point[i] > ih[i]->common.ndrec) {
			Fprintf(stderr,
			 "mux: end point %d greater than ndrec for file %s",
			 e_point[i], in_files[i]);
			Fprintf(stderr,", reset to %d\n",ih[i]->common.ndrec);
		        e_point[i] = ih[i]->common.ndrec;
		}

		start_time_offset = (s_point[i] - 1) / first_sf;

		if (genhd_size == 1)
			for (j = 0; j < num_channels; j++)
				start_times[start_time_chan++] = 
					*start_time_file + start_time_offset;
		else
			for (j = 0; j < num_channels; j++)
				start_times[start_time_chan++] =
					start_time_file[j] + start_time_offset;

		switch (genhd_type("max_value", &genhd_size, ih[i])) {
		case HD_UNDEF:
			no_max_value = YES;
			break;
		case DOUBLE:
			if (genhd_size != num_channels && genhd_size != 1)
			{
				Fprintf(stderr,
				 "mux: header item \"max_value\" in file %s has wrong size (%d).\n",
					in_files[i], genhd_size);
				exit(1);
			}
			break;
		default:
			Fprintf(stderr,
			 "mux: header item \"max_value\" in file %s is not DOUBLE.\n",
				in_files[i]);
		}

		if (!no_max_value)
		{
			max_value_file = get_genhd_d("max_value", ih[i]);
			if (genhd_size == 1)
				for (j = 0; j < num_channels; j++)
					max_values[max_value_chan++] =
						*max_value_file;
			else
				for (j = 0; j < num_channels; j++)
					max_values[max_value_chan++] =
						max_value_file[j];
		}

		
/* skip to the first record to be processed */
		if (s_point[i]-1)
			fea_skiprec(in_fd[i], s_point[i]-1, ih[i]);
	}


	if (Jflag) {
		int k=0;
		for(j=0; j<total_channels; j=j+2) {
	   	   if (start_times[j] != start_times[j+1]) {
			Fprintf(stderr,
			 "mux: Warning, start times of channels to be joined inconsistent,\n");
			Fprintf(stderr,
			 "mux: start time for first channel (real part) used.\n");
		   }
		start_times[k++]=start_times[j];
		}
		total_channels /= 2;
	}

	if (debug_level) {
		Fprintf(stderr, "Total Channels: %d\n", total_channels);
		Fprintf(stderr, "First_sf: %g", first_sf);
		Fprintf(stderr, " First_type: %s\n", type_codes[first_type]);
		for(i=0;i<nfiles;i++) {
			Fprintf(stderr,"s_point[%d]: %d, e_point[%d]: %d, ",
		 	i,s_point[i],i,e_point[i]);
		}
		for(i=0;i<total_channels;i++)
			Fprintf(stderr,"start_times[%d]: %g\n",
		 	i,start_times[i]);
	}

	if (!Jflag)
		out_type = first_type;
	else
		switch (first_type) {
		case DOUBLE:
			out_type = DOUBLE_CPLX;
			break;
		case FLOAT:
			out_type = FLOAT_CPLX;
			break;
		case LONG:
			out_type = LONG_CPLX;
			break;
		case SHORT:
			out_type = SHORT_CPLX;
			break;
		case BYTE:
		case CHAR:
			out_type = BYTE_CPLX;
			break;
		}

	start_time_tmp = start_times[0];
	multi_start_time = NO;
	for (i=1; i<total_channels; i++) 
		if (start_time_tmp != start_times[i]) {
			multi_start_time = YES;
			break;
		}

	if(init_feasd_hd(oh, out_type, total_channels, start_times,
			 multi_start_time, first_sf)) {
		ERROR_EXIT("error creating output header");
	}

	if(!no_max_value) 
		(void) add_genhd_d("max_value", max_values, total_channels, oh);
	write_header(oh, out_fd);

/* allocate output and input records */
	
	out_rec = allo_feasd_recs(oh, out_type, 1L, (char *) NULL, NO);
	out_data = out_rec->data;

	in_rec = (struct feasd **) calloc((unsigned) nfiles, 
					  sizeof(struct feasd *));
	if (Jflag)
	{
		in_data = malloc(2*total_channels*typesiz(first_type));
		spsassert(in_data, "mux: malloc failed.");
	}
	else	in_data = out_data;

	prev_num_chans = 0;
	for (i=0; i<nfiles; i++) {
		in_rec[i] = allo_feasd_recs(ih[i], first_type, 1L,
			in_data + typesiz(first_type)*prev_num_chans,
			NO);
		prev_num_chans += get_chans(ih[i]);
	}

/* read-write loop */

	while (!files_done) {
		for (i=0; i<nfiles; i++) {
		    if(get_feasd_recs(in_rec[i], 0L, 1L, ih[i], in_fd[i]) == 0)
			files_done = 1;
		    if(n_read[i]++ > e_point[i]-s_point[i])
			files_done = 1;
		}

		if (!files_done)
		{
		    if (Jflag)
		    {
#define			CASE(TYPE, real_t, cplx_t)			\
			case TYPE:					\
			    {						\
				real_t *in = (real_t *) in_data;	\
				cplx_t *out = (cplx_t *) out_data;	\
									\
				for (j = 0; j < total_channels; j++)	\
				{					\
				    out[j].real = *in++;		\
				    out[j].imag = *in++;		\
				}					\
			    }						\
			    break

			switch (first_type)
			{
			CASE(DOUBLE, double, double_cplx);
			CASE(FLOAT, float, float_cplx);
			CASE(LONG, long, long_cplx);
			CASE(SHORT, short, short_cplx);
			case CHAR:	/* FALL THROUGH */
			CASE(BYTE, char, byte_cplx);
			}

#undef			CASE
		    }

		    if(put_feasd_recs(out_rec,0L,1L,oh,out_fd)!=0) {
			ERROR_EXIT("error writing output file");
		    }

		    output_count++;
		}
	}


/*
 * put output file info in ESPS common
 */
    	if (strcmp(ofile, "<stdout>") != 0) {
        	(void)putsym_s("filename",in_files[0]);
        	(void)putsym_s("prog",ProgName);
        	(void)putsym_i("start",1);
        	(void)putsym_i("nan",(int) output_count);
    	}
    	if (debug_level) Fprintf(stderr, "mux: normal exit\n");

	exit(0);
	/*NOTREACHED*/
}


long
get_chans(hd)
struct header *hd;
{
	return get_fea_siz("samples",hd,(short *)NULL, (long **)NULL);
}
